src/garage/_primitives.py

```diff
@@ -0,0 +1,152 @@
+"""Policy interface."""
+import abc
+
+
+class Primitive(abc.ABC):
+    """Primitive interface.
+
+    Args:
+        name (str): Name for logs, variable scopes, etc.
+        spec (garage.InOutSpec): Input/output specification this primitive will
+            satisfy.
+
+    """
+
+    def __init__(self, name, spec):
+        self._name = name
+        self._spec = spec
+
+    @abc.abstractmethod
+    def get_parameters(self):
+        """Get parameters of this primitive.
+
+        Returns:
+            dict[str, np.ndarray]: A dictionary mapping from names to values of
+                each parameter.
+
+        """
+
+    @abc.abstractmethod
+    def set_parameters(self, parameters):
+        """Set the parameters of this primtive.
+
+        Args:
+            parameters (dict[str, np.ndarray]): A dictionary mapping from names
+                to values of each parameter.
+
+        """
+
+    @property
+    def spec(self):
```

Comment:
I think `InOutSpec` is a good argument to pass into this API, but I'm not convinced it should actually be accessible.

Reply:
@krzentner Without this property, how can it be connected with other primitives? Given that the spec should be compatible between two primitives when they are connected.

Reply:
It's true that `InOutSpec` is useful if I have two things with arbitrary shapes (e.g. a pre-trained model and an environment that model wasn't trained on), and I want to create a model to automatically connect them together. However, we've very rarely wanted to do that in practice.

I'm not saying this property is a bad idea, but it's much easier to add new properties later than to remove existing ones, and it's not clear this property is necessary in most cases. In the rare cases that we do need it, I think it would be fine to put this in a separate interface.

If the `spec` property is supposed to also have the same semantics as an environment's `spec` property, then my opinion that this should be in a separate interface is even stronger.

Reply:
So I think this really gets at the heart of what we are going to ask of models in garage.

BLUF: Do we want to make safe runtime model composition first-class or not? If we do, then I think we need `.spec` interface, but maybe I'm missing something.

By "runtime" I mean that someone can define higher-order models like `ContextConditionedPolicy(ContextEncoderModel, PolicyModel)` which don't require users to pass classes+args to their constructors, but instead accept instances and wire themselves using the `.spec` provided by the instance.

This proposal basically says a model:
1. Has a name (discussion in another thread)
2. Exposes zero-or-more parameters, which I can get/set
3. Can be "wired together" with other models safely at runtime, using specs as a runtime type checking system
  a. By validating an InOutSpec in its constructor
  b. By providing its own InOutSpec, allowing the "wirer" to generate a derived InOutSpec

IME, if we are going to have 3a, not having 3b can make things really difficult.

I also don't agree that we don't really have this problem yet. It already appears
in PEARL: https://github.com/rlworkgroup/garage/blob/master/examples/torch/pearl_ml1_push.py
in Task Embeddings: 
https://github.com/rlworkgroup/garage/pull/1171/files#diff-47f13f3e683bface4baa8af5f8d921c4R127
..and the current state makes things really awkward to implement, and encourages creating one-off monolithic models for simple compositions (e.g. `GaussianMLPMultitaskPolicy`).

Reply:
I think you have the APIs needed for 3a and 3b backwards from what I expect. Personally, I think we should prioritize 3b (generating derived `InOutSpec` for sub-models) over 3a (checking that all the `InOutSpec` we've been given look right).

Consider the example you give above: `ContextConditionedPolicy(ContextEncoderModel, PolicyModel)`. I think to make this example "just work", `ContextConditionedPolicy` should be generating the `InOutSpec` for `ContextEncoderModel` and `PolicyModel` from its own `InOutSpec` (which is just the transposed environment spec), and passing those generated specs down to those sub-models. Basically, specs should be processed in an "outside-in" order. That's the obvious way to ensure everything fits together (i.e. 3b). No part of that process (except for the environment itself) involves reading `.spec`, only passing it down as a constructor arg.

Alternatively, you can check that `.spec`s fit together, but not design the API to ensure they do from the start. That, to me, sounds unwise. Although it might allow for slightly better error messages, the checking is not meaningfully earlier than just running the resulting model on real data. Furthermore, if `.spec` doesn't match the actual data handled / produced by the model, then you may end up with false negatives or false positives in you checking.

It might also make sense to read a `.spec` field if you're given two things that *don't* fit together, and you're going to generate another model to glue them together. Personally, I think it would be weird for `ContextConditionedPolicy` to construct additional trainable matrices to fit its sub-models together, so I don't think it even needs to read `.spec`. It might make sense for e.g. PEARL itself to do that though.

I agree that some cases may benefit from having the `.spec` interface, and I definitely think we're going to have a `.spec` interface one way or another. The question I'm trying to get at really is if third-party author's models need to implement it, when we can't even come up with a good example of when we would need to read such a field.

If we bundle it in with the `Model` interface, then the answer appears to be that they must (and environments simply happen to implement an identical but unrelated interface). If it's separate, then when it's needed the function can say so explicitly.